aptos-move/mvhashmap/src/versioned_group_data.rs

```diff
@@ -78,42 +81,90 @@ impl<T: Hash + Clone + Debug + Eq + Serialize, V: TransactionWrite> Default
 }
 
 impl<T: Hash + Clone + Debug + Eq + Serialize, V: TransactionWrite> VersionedGroupValue<T, V> {
-    fn set_base_values(
-        &mut self,
-        shifted_idx: ShiftedTxnIndex,
-        // TODO[agg_v2](fix) add layout to values
-        values: impl IntoIterator<Item = (T, V)>,
-    ) {
+    fn set_raw_base_values(&mut self, values: impl Iterator<Item = (T, V)>) {
+        let shifted_idx = ShiftedTxnIndex::zero();
         match self.idx_to_update.get(&shifted_idx) {
             Some(previous) => {
                 // base value may have already been provided by another transaction
                 // executed simultaneously and asking for the same resource group.
                 // Value from storage must be identical, but then delayed field
                 // identifier exchange could've modified it.
                 //
-                // If maybe_layout is None, they are required to be identical
-                // If maybe_layout is Some, there might have been an exchange
+                // If they are RawFromStorage, they need to be identical.
+                //
                 // Assert the length of bytes for efficiency (instead of full equality)
-                for (tag, v) in values.into_iter() {
+                for (tag, v) in values {
                     let prev_v = previous
                         .get(&tag)
                         .expect("Reading twice from storage must be consistent");
-                    assert!(v.bytes().map(|b| b.len()) == prev_v.bytes().map(|b| b.len()));
+                    if let ValueWithLayout::RawFromStorage(prev_v) = prev_v {
+                        assert!(v.bytes().map(|b| b.len()) == prev_v.bytes().map(|b| b.len()));
+                    }
                 }
             },
             // For base value, incarnation is irrelevant, and is always set to 0.
             None => {
-                self.write(shifted_idx, 0, values);
+                self.write(
+                    shifted_idx,
+                    0,
+                    values.map(|(k, v)| (k, ValueWithLayout::RawFromStorage(Arc::new(v)))),
+                );
             },
         }
     }
 
+    fn update_tagged_base_value_with_layout(
+        &mut self,
+        tag: T,
+        value: V,
+        layout: Option<Arc<MoveTypeLayout>>,
+    ) {
+        let shifted_idx = ShiftedTxnIndex::zero();
+        let v = ValueWithLayout::Exchanged(Arc::new(value), layout.clone());
+
+        use btree_map::Entry::*;
+        match self
+            .versioned_map
+            .entry(tag.clone())
+            .or_default()
+            .entry(shifted_idx.clone())
```

Comment:
minor: why not get_mut().expect("Base version must exist when updating for exchange") in this case, since vacant is unreachable? (I typically like whatever makes it less indented)

Reply:
When I tried using get_mut().expect(".."), the call "o.insert(..)" raises an error that the insert function doesn't exist. May be I am not doing it the right way.

Reply:
That's because o will no longer be an entry, but an actual reference to value, insert was the method of OccupiedEntry (https://doc.rust-lang.org/beta/std/collections/btree_map/struct.OccupiedEntry.html#method.insert), but now you can directly assign.

Reply:
let's leave like this until we figure out if we are comfortable with the panic, or need to raise the PanicError upstream?